### PR TITLE
[bugfix] Encoding issue for the Test Runner

### LINKED ISSUE #659: Accessing a resource scoped configuration without providing a resource is not expected
### Version:
```
1. VSCode: 1.32.3
2. JDK: 1.8.0_202
3. Extension: 
- Langauge Support for Java(TM) by Red Hat: 0.41.0
- Debugger for Java: 0.17.0
- Java Test Runner: 0.15.1
```

### Repro steps:

1. Open a java project in vscode
2. Trigger any test from test explorer or codelens

### Result:
image

### PR DESCRIPTION
**Pull request by jdneo**
encoding

Will also fix #659

### CODE REVIEW

File: src/runners/baseRunner/BaseRunner.ts [hunk 1]
```diff
@@ -203,9 +206,33 @@ export abstract class BaseRunner implements ITestRunner {
...
+                }
+            }
+        }
+        return undefined;
+    }
+
+    private getEncodingFromSetting(): string {
         let javaEncoding: string | null = null;
+        // One runner will contain all tests in one workspace with the same test framework.
+        const config: WorkspaceConfiguration = workspace.getConfiguration(undefined, Uri.parse(this.tests[0].location.uri));
```
Comment (yaohaizh) [inline src/runners/baseRunner/BaseRunner.ts]: Why get all configuration?

File: src/runners/baseRunner/BaseRunner.ts [hunk 2]
```diff
@@ -62,29 +65,29 @@ export abstract class BaseRunner implements ITestRunner {
...
-                buffer = buffer.concat(data.toString());
-                const index: number = buffer.lastIndexOf(os.EOL);
+            this.process.stdout.on('data', (buffer: Buffer) => {
+                data = data.concat(iconv.decode(buffer, this.encoding));
+                const index: number = data.lastIndexOf(os.EOL);
                 if (index >= 0) {
-                    testResultAnalyzer.analyzeData(buffer.substring(0, index));
-                    buffer = buffer.substring(index);
+                    testResultAnalyzer.analyzeData(data.substring(0, index));
+                    data = data.substring(index);
```
Comment (yaohaizh) [inline src/runners/baseRunner/BaseRunner.ts]: Should bypass os.EOL?

File: src/runners/baseRunner/BaseRunner.ts [hunk 3]
```diff
@@ -203,9 +206,33 @@ export abstract class BaseRunner implements ITestRunner {
...
+            throw new Error(`Invalid encoding: ${encoding}`);
+        }
+        return encoding;
+    }
+
+    private getEncodingFromTestConfig(): string | undefined {
+        const encodingArgPrefix: string = '-Dfile.encoding=';
+        if (this.config && this.config.vmargs) {
+            // loop backwards since the latter vmarg will override the previous one
+            for (let i: number = this.config.vmargs.length - 1; i >= 0; i--) {
```
Comment (yaohaizh) [inline src/runners/baseRunner/BaseRunner.ts]: let ==> const

File: src/runners/baseRunner/BaseRunner.ts [hunk 4]
```diff
@@ -203,9 +206,33 @@ export abstract class BaseRunner implements ITestRunner {
...
+        const encoding: string = this.getEncodingFromTestConfig() || this.getEncodingFromSetting();
+        if (!iconv.encodingExists(encoding)) {
+            throw new Error(`Invalid encoding: ${encoding}`);
+        }
+        return encoding;
+    }
+
+    private getEncodingFromTestConfig(): string | undefined {
+        const encodingArgPrefix: string = '-Dfile.encoding=';
+        if (this.config && this.config.vmargs) {
```
Comment (testforstephen) [inline src/runners/baseRunner/BaseRunner.ts]: Use regular expression to shorten the logic.

**Pull request merged**